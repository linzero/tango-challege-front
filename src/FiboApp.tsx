import { useDispatch, useSelector } from "react-redux";
import { getNumber, setNumber } from "./actions/fiboActions";
import { fiboState } from "./reducers/fiboReducer";
import { uiState } from "./reducers/uiReducer";

interface storeState {
    fibo:fiboState
    ui:uiState
}

const FiboApp = () => {

    const dispatch = useDispatch();
    const { fibo:{response}, ui:{loading}, fibo:{num} } = useSelector( (state:storeState) => state );

    const handleInputChange = (e:any) => {
        dispatch( setNumber( e.target.value ) )
    }

    const handleButtonClic = () => {
        dispatch( getNumber( num || 0 ) )
    }

    return (
        <>
            <div className="main-content d-flex align-items-center justify-content-center">
                <div className="container text-center">
                    <h1 className="mb-3">Calculate Fibonacci!</h1>
                    <form className="main-form row m-0 m-auto mb-3">
                        <div className="col-6">
                            <input className="form-control" onChange={handleInputChange} value={num} />
                        </div>
                        <div className="col-6">
                        {
                            (!loading) && <button className="btn btn-primary w-100" onClick={handleButtonClic}>Calculate</button>
                        }
                        {
                            (loading) && <button className="btn btn-primary w-100" disabled>Calculando...</button>
                        }
                        </div>
                    </form>

                    {
                        (!( isNaN(parseInt(response?.toString() || '')) )) && 
                        <div>
                            <p><span className="middle"><b>Fibonacci Result:</b></span> <span id="content-response" className="fs-4 middle">{ response }</span></p>
                        </div>
                    }
                </div>
            </div>
            
        </>
    )
}

FiboApp.propTypes = {

}

export default FiboApp
