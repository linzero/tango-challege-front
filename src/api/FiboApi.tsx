import axios from "axios";
import { enviroment } from "../enviroment/enviroment";

export const getNumber_GET = async (num:number) => {
    return axios.get(`${ enviroment.apiBaseUrl }/${num}`)
}