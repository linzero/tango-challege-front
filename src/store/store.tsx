import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import { fiboReducer } from "../reducers/fiboReducer";
import thunk from 'redux-thunk';
import { uiReducer } from "../reducers/uiReducer";

const win:any = window
const composeEnhancers = (typeof win !== 'undefined' && win.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const reducers = combineReducers({
    fibo:fiboReducer,
    ui:uiReducer
});

export const store = createStore(
    reducers,
    composeEnhancers(
        applyMiddleware( thunk )
    )
);