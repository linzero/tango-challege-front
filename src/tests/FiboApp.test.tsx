import { mount } from "enzyme"
import { Provider } from "react-redux"
import { getNumber_GET } from "../api/FiboApi"
import FiboApp from "../FiboApp"
import { store } from "../store/store"

export default describe('Fibonacci Test', (initialState={}) => {
    test( 'First Display <FiboApp></FiboApp>',() =>{

        const wrapper = mount( <Provider store={ store }><FiboApp /></Provider> )

        expect( wrapper ).toMatchSnapshot();
    })

    test( 'API Test', async() =>{

        try {
            const values = [0,1,1,2,3,5,8]

            for (let i = 0; i<values.length; i++ ) {
                const result = await getNumber_GET(i)

                expect(result.data.response).toBe(values[i])
            }

        }
        catch(e:any){
            throw new Error(e)
        }

    })
})