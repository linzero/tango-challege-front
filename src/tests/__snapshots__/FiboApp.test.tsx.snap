// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fibonacci Test First Display <FiboApp></FiboApp> 1`] = `
<div
  className="main-content d-flex align-items-center justify-content-center"
>
  <div
    className="container text-center"
  >
    <h1
      className="mb-3"
    >
      Calculate Fibonacci!
    </h1>
    <form
      className="main-form row m-0 m-auto mb-3"
    >
      <div
        className="col-6"
      >
        <input
          className="form-control"
          onChange={[Function]}
          value={0}
        />
      </div>
      <div
        className="col-6"
      >
        <button
          className="btn btn-primary w-100"
          onClick={[Function]}
        >
          Calculate
        </button>
      </div>
    </form>
  </div>
</div>
`;
